core/src/main/java/hivemall/mf/CofactorizationUDTF.java

```diff
@@ -0,0 +1,574 @@
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.StringObjectInspector;
+import org.apache.hadoop.mapred.Counters;
+import org.apache.hadoop.mapred.Reporter;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import static hivemall.utils.lang.Primitives.FALSE_BYTE;
+import static hivemall.utils.lang.Primitives.TRUE_BYTE;
+
+public class CofactorizationUDTF extends UDTFWithOptions {
+    private static final Log LOG = LogFactory.getLog(CofactorizationUDTF.class);
+
+    // Option variables
+    // The number of latent factors
+    protected int factor;
+    // The scaling hyperparameter for zero entries in the rank matrix
+    protected float scale_zero;
+    // The scaling hyperparameter for non-zero entries in the rank matrix
+    protected float scale_nonzero;
+    // The preferred size of the miniBatch for training
+    protected int batchSize;
+    // The initial mean rating
+    protected float globalBias;
+    // Whether update (and return) the mean rating or not
+    protected boolean updateGlobalBias;
+    // The number of iterations
+    protected int maxIters;
+    // Whether to use bias clause
+    protected boolean useBiasClause;
+    // Whether to use normalization
+    protected boolean useL2Norm;
+    // regularization hyperparameters
+    protected float lambdaTheta;
+    protected float lambdaBeta;
+    protected float lambdaGamma;
+
+    // Initialization strategy of rank matrix
+    protected CofactorModel.RankInitScheme rankInit;
+
+    // Model itself
+    protected CofactorModel model;
+    protected int numItems;
+
+    // Variable managing status of learning
+
+    // The number of processed training examples
+    protected long count;
+
+    protected ConversionState cvState;
+    private ConversionState validationState;
+
+    // Input OIs and Context
+    protected StringObjectInspector contextOI;
+    protected ListObjectInspector featuresOI;
+    protected BooleanObjectInspector isItemOI;
+    protected ListObjectInspector sppmiOI;
+
+    // Used for iterations
+    protected NioStatefulSegment fileIO;
+    protected ByteBuffer inputBuf;
+    private long lastWritePos;
+
+    private Feature contextProbe;
+    private Feature[] featuresProbe;
+    private Feature[] sppmiProbe;
+    private boolean isItemProbe;
+    private long numValidations;
+    private long numTraining;
+
+
+    static class MiniBatch {
+        protected int maxSize;
+        private List<TrainingSample> users;
+        private List<TrainingSample> items;
+
+        protected MiniBatch(int maxSize) {
+            this.maxSize = maxSize;
+        }
+
+        protected void add(TrainingSample sample) {
+            if (size() == this.maxSize) {
+                return;
```

Comment:
`sample` is thrown away?

You can return boolean value to check MiniBatch is full or not.

Reply:
I think an explicit mini batch size is not required anymore. The upper limit is the size of the `inputBuf`. Is this correct?